### DISCUSSION TITLE
Packaging

### DISCUSSION THREAD

**Discussion by Leon0402** [friendlyanon/cmake-init#21] [Q&A]
I think a really nice thing a lot of people struggle with would be to help a little bit with packaging. What you think?

**Comment by friendlyanon** [comment]
Packaging is the kind of thing where you have extreme combinatorial explosion of options, so I can't really do anything in that regard. Just as an example, it's [non-trivial](https://github.com/friendlyanon/cxx-static-shared-example/tree/debian-packaging/cmake/packaging/deb) to produce a `.deb` package with CPack that passes pedantic lintian checks.

**Reply by Leon0402** [reply]
Yes I know it's a difficult and I would certainly not create anything by default here. But especially because it's quite difficult it would be of much benefit to auto generate a few of the options! 

My general standpoint would be: Do not add anything by default most people don't need and don't make the init script super complex. 
But something like: 
"Do you want packaging": Y  -> If no, skip
"Do you target Linux":  Y
"Need .deb packaging": Y -> Generate non trivial Code

No need to make add every possible generator and option and make this extremly complex. I

**Reply by friendlyanon** [reply]
Anything that this project would include would only work for a handful of people. You also have to consider that pretty much everything is cross-platform in cmake-init generated projects, with the exception of a few worthy inclusions that are generally useful, like sanitizers and coverage, although for coverage you only have an example for OpenCppCoverage included for those who wish to develop something Windows-only using cmake-init. These are things that everyone should use at all times, however, so the priority is quite high to have these included.  
I consider fuzzing to be an equally important thing and that's why there is an example repository showing fuzzing for a non-toy project, but it's not generally useful, only when you are developing a parser, so it's no included by default.

Packaging is not something a developer in general has to do. There are package maintainers for that! The readme even says that cmake-init generated projects are structured in a way that makes them trivially packageable. This is why there is a separate developer mode. Once a project gets **any** amount of users (including package maintainers), the developer will be in the minority.

**Reply by Leon0402** [reply]
hmm you mean by "package manager" people from the distros that do such stuff? Like people writing AUR scripts for arch or debian maintainers creating a deb? 

Yeah that's possible if you have a well known software and in the best case it should be free or at leats open source software as well for say debian. 
I don't think this holds true for most cmake projects. 

Loads of software provide their own .deb files for instance. And for say Windows or Mac there is nobody who will do the job for you. I don't know, I think a setup, which allows me to bring my app into the AppStore would be nice. 
Even if it just people need it, who create an Application and target Mac OS need it. Why is that a argument for not generating it optionally? I consider this the advantage of an init script in contrast to a static template. 
Or is it the maintaince overhead? And you say it's not worth it?

**Reply by friendlyanon** [reply]
For Windows, you have Chocolatey picking up some projects and for Mac there is Homebrew picking up a lot more projects. Same applies to different distros having different packaging formats and distribution methods.  
But these methods are also package formats on their own, further proving that there really isn't a general solution here.

There are also dependencies. How does each platform deal with those? How are they built, as static or shared? Where are they stored, `lib`, `lib64` or somewhere else? Do the dependencies require patches to work on platforms? What about interactions with other applications?  
These are non-trivial questions I could come up with just for dependencies alone.

You say that `.deb` is a commonly provided format, but you probably saw what it takes to build a proper `.deb` and it's not pretty. Technically, [it's UB](https://github.com/friendlyanon/cxx-static-shared-example/blob/debian-packaging/cmake/packaging/deb/config.cmake#L85-L89) even.  
CPack missed out on some of the modernizations that other components of CMake received recently.

**Comment by friendlyanon** [comment]
One of the [examples](https://github.com/friendlyanon/cmake-init/wiki/Examples) shows how to do packaging: https://github.com/friendlyanon/cmake-init-multiple-packages

This one shows how to solve a particular problem and I will make more as I come across such problems. If you have more in mind, then you could list them.
